Build from source prototype demo scenario (#862)
pyrsia/pyrsia: docs/tutorials/build_from_source.md

# How to setup a Pyrsia node to build from source

> **Warning:** The build-from-source demo is still work-in-progress.

This tutorial describes how to setup a Pyrsia node that can build artifacts from
source with the goal to publish them in the Pyrsia network.

Ultimately, the following scenario will be used, but for now some steps
(indicated below) are skipped for the purpose of this build-from-source demo:

- Setup at least 3 authorized nodes (skipped in this demo, only one Pyrsia node
  is used)
- Make sure a mapping between an artifact and its source exists in the
  [Pyrsia Mappings Repo](https://github.com/pyrsia/pyrsia-mappings) (for Maven artifacts
  only)
- Set up a build pipeline per Pyrsia node and configure the nodes to use them
- Trigger a build from source for a given artifact
- Wait for the build to finish in the build pipeline
- Try to reach consensus with the other authorized nodes, which have to run the

>   same build and verify they produce the same build result. (skipped in this demo)

Reviewer: `s/skipped/Skipped`

- Create a transparency log about the artifact publication
- Publish the artifact on the p2p network

See the [architecture and use-cases](../developers/pyrsia-architecture-and-use-cases.md)
document for more information.

Because this demo scenario results in a published Maven artifact in the Pyrsia
network, we can run a final step to show the build from source worked:

> - Use Pyrsia in a Maven project

Reviewer: Confusing this bullet point here, maybe move it up with the others: `- [Optional] Use Pyrsia...`

Author: it's technically not part of the build-from-source which is why I would like to keep it separate

## Compile Pyrsia

Download a fresh copy of the codebase by cloning the repo or updating to the
HEAD of `main`.

```sh

> git clone git@github.com:pyrsia/pyrsia.git

Reviewer 2: Use the public https url?

```

Let's call this folder `PYRSIA_HOME`. We will refer to this
name in the following steps.

> Build binaries for by running:

Reviewer: remove `for`?

> Build binaries for by running:

Reviewer 2: ```suggestion
Build binaries for pyrsia by running:
```

```sh
cd $PYRSIA_HOME
cargo build --workspace
```

## Run the Pyrsia node

Run the Pyrsia node using `cargo run`. We set the following env vars:

- RUST_LOG: to make sure we can see all the debug logs
- DEV_MODE: to make sure all non-existing directories are created on-the-fly

> - PYRSIA_ARTIFACT_PATH: to point to a clean space to store artifacts so we can

Author: I've rephrase

  easily start from scratch

```sh
RUST_LOG=pyrsia=debug DEV_MODE=on PYRSIA_ARTIFACT_PATH=/tmp/pyrsia \
cargo run --package pyrsia_node -- --pipeline-service-endpoint  http://localhost:8080
```

> As you can see we specified the `--pipeline service endpoint` argument to point

Reviewer 2: ```suggestion
As you can see we specified the `--pipeline-service-endpoint` argument to point
```

to `http://localhost:8080`, which is where we will run our build pipeline prototype

> (see below). In a real setup, the build pipeline obviously needs to run on its

Reviewer 2: ```suggestion
(see below). In a production setup, the build pipeline needs to run on its
```

artifact, this repository has a mapping file like this:

```json
{
  "package_type": "Maven2",
  "package_specific_id": "groupId:artifactId:version",
  "source_repository": {
    "Git": {
      "url": "https://github.com/organization/artifact",
      "tag": "rel/artifact-version-tag"
    }
  },
  "build_spec_url": ""
}
```

In case you want to build a different artifact from source than already available
in the mapping, feel free to create a pull request to add it to the mapping repository.

In this demo we will build `commons-codec:commons-codec:1.15` for which the mapping
is already available.

## Run build pipeline prototype

In this demo, we use a build pipeline prototype. Eventually, Pyrsia will use
industry-standard build pipeline solutions. Integration with such solutions is
currently being investigated.

The most important features of the build pipeline are:

- it runs completely separate from the Pyrsia node
- it exposes an interface so Pyrsia can start a build
- it exposes an interface so Pyrsia can download the build output

The current build pipeline prototype only supports Maven builds. It takes

> the artifact mapping as input (see above), start a Maven build and provide the

Reviewer 2: ```suggestion
the artifact mapping as input (see above), starts a Maven build and provides the
```

> the artifact mapping as input (see above), start a Maven build and provide the

Reviewer: `s/start/starts`, `s/provide/provides`

build output as a download.

Download or clone the [prototype repo](https://github.com/tiainen/pyrsia_build_pipeline_prototype)
and run as follows:

```sh
cd pyrsia_build_pipeline_prototype
RUST_LOG=debug cargo run
```

By default, this prototype listens on http port 8080. If you run it on a different

> host or port, make sure the specify its location when starting the Pyrsia node

Reviewer 2: ```suggestion
host or port, make sure to specify its location when starting the Pyrsia node
```

> with `--pipeline service endpoint` (see above).

Reviewer 2: ```suggestion
with `--pipeline-service-endpoint` (see above).
```

  "operation":"AddArtifact"
}
```

## Publish the artifact on the p2p network

As a final step in the build from source scenario, the artifacts are stored locally
and provided on the p2p network.

```text
 INFO  pyrsia::artifact_service::service > put_artifact with id: da341557-9150-4208-9474-f5884f799338
 INFO  pyrsia::artifact_service::storage > An artifact is being pushed to the artifact manager da341557-9150-4208-9474-f5884f799338
 DEBUG pyrsia::network::client           > p2p::Client::provide "da341557-9150-4208-9474-f5884f799338"
 ```

## Use Pyrsia in a Maven project

Now that we have a published Maven artifact in the Pyrsia network, we can try to
use it in a Maven project.

Create a Java project:

```sh
mkdir pyrsia-maven-sample
cd pyrsia-maven-sample
mkdir -p src/main/java/org/pyrsia/sample
```

Create a file `src/main/java/org/pyrsia/sample/Main.java`:

```java
package org.pyrsia.sample;

import java.util.Arrays;
import org.apache.commons.codec.binary.Hex;
import static java.lang.System.out;

public class Main {

    public static void main(String[] args) {
        byte[] data = { 1, 2, 3, 4, 5, 6, 7, 8 };
        String hexEncodedData = Hex.encodeHexString(data);

>         out.println("Byte array " + Arrays.toString(data) + " encoded as a hex string: " + hexEncodedData);

Reviewer: Use `System.out.println`

    }
}
```

The code in this sample uses `org.apache.commons.codec.binary.Hex` from the commons-codec
library, so let's add this as dependency in our Maven build:

Create a `pom.xml` file:

```xml
<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xmlns="http://maven.apache.org/POM/4.0.0"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
              http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>org.pyrsia.sample</groupId>
    <artifactId>maven</artifactId>
    <packaging>jar</packaging>
    <version>1.0.0-SNAPSHOT</version>
    <name>Pyrsia Sample for Maven</name>

    <repositories>
        <repository>
            <id>pyrsia</id>
            <url>http://localhost:7888/maven2</url>
        </repository>
    </repositories>

    <dependencies>
        <dependency>
            <groupId>commons-codec</groupId>
            <artifactId>commons-codec</artifactId>
            <version>1.15</version>
        </dependency>
    </dependencies>
</project>
```

As you can see we have set the repository to `http://localhost:7888/maven2`, which

> will trigger to Maven build to request dependencies from this URL, which is our

Reviewer 2: ```suggestion
will trigger the Maven build to request dependencies from this URL, which is our
```

[INFO] Total time:  0.925 s
[INFO] Finished at: 2022-07-15T16:33:39+02:00
[INFO] ------------------------------------------------------------------------
```

The dependency was downloaded from Pyrsia, which you can verify in the Pyrsia node
logs:

```text
 DEBUG pyrsia::java::maven2::routes                    > route full path: /maven2/commons-codec/commons-codec/1.15/commons-codec-1.15.jar
 DEBUG pyrsia::java::maven2::handlers::maven_artifacts > Requesting maven artifact: /maven2/commons-codec/commons-codec/1.15/commons-codec-1.15.jar
 DEBUG pyrsia::java::maven2::handlers::maven_artifacts > Requesting artifact for id commons-codec/commons-codec/1.15/commons-codec-1.15.jar
 INFO  pyrsia::artifact_service::storage               > An artifact is being pulled from the artifact manager a90e6792-4f2e-4ccc-8479-d935431e28ec
 DEBUG pyrsia::artifact_service::storage               > Pulling artifact from /private/tmp/pyrsia/a90e6792-4f2e-4ccc-8479-d935431e28ec.file
 INFO  pyrsia_registry                                 > 127.0.0.1:55273 "GET /maven2/commons-codec/commons-codec/1.15/commons-codec-1.15.jar HTTP/1.1" 200 "-" "Apache-Maven/3.8.5 (Java 1.8.0_332; Mac OS X 12.4)" 23.00275ms

> ```

Reviewer: Since there is a demo, it would be nice to run it.
Something like:
```sh
$ java -cp target/classes:/Users/$user/.m2/repository/commons-codec/commons-codec/1.15/commons-codec-1.15.jar org.pyrsia.sample.Main
Byte array [1, 2, 3, 4, 5, 6, 7, 8] encoded as a hex string: 0102030405060708
```
(`java -jar /target/maven-1.0.0-SNAPSHOT.jar org.pyrsia.sample.Main` doesn't work because a main attribute is missing in the manifest, and solving it would require a plugin in the pom)

Author: this adds too much (Java) complexity that isn't very useful in this demo. For me the demo ends when the artifact is downloaded.

Running the sample would maybe better fit a shared java sample that we commit somewhere
